### ISSUE TITLE
🚀 Question: Is attention_mask still lower triangular when using packing=True in SFTTrainer?

### ISSUE THREAD

**Issue by noforit** [huggingface/trl#3067]
Hello,

I am using SFTTrainer with `packing=True` to pack multiple samples into a single sequence:

```
trainer = SFTTrainer(
    model=model,
    args=training_args,
    train_dataset=train_data,
    eval_dataset=val_data,
    peft_config=lora_config,
    packing=True,
)
```

I have a question regarding the behavior of attention_mask in this scenario:
When multiple samples are packed together into one sequence, is the attention_mask still lower triangular?
If not, can flash_attn2 still achieve acceleration in this case?
Looking forward to your insights. Thanks in advance! 🚀

**Comment by qgallouedec** [comment]
This may help you #3076

**Comment by qgallouedec** [comment]
The answer is yes and no. It's still triangular but the samples can "contaminate". That's a known issue #1230

**Comment by qgallouedec** [comment]
Update answer: with the default `packing_strategy="bfd"` (BFD), TRL now auto-enables padding-free mode and passes `position_ids` that segment the packed sequences, so FlashAttention-2 attends within each sample only and you keep the speedup without cross-contamination.

With the legacy `packing_strategy="wrapped"` the mask is still causal but samples can contaminate each other, which is the behavior I described originally. Since the question is answered and the underlying concern is solved on current main, I'll close this.
